sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/MutationGroupEncoder.java

```diff
@@ -0,0 +1,649 @@
+    List<String> allTables = schema.getTables();
+    String table = m.getTable().toLowerCase();
+    int tableIndex = allTables.indexOf(table);
+    VarInt.encode(tableIndex, bos);
+    ObjectOutput out = new ObjectOutputStream(bos);
+    out.writeObject(m.getKeySet());
+  }
+
+  private Mutation decodeDelete(ByteArrayInputStream bis)
+      throws IOException {
+    List<String> tables = schema.getTables();
+    int tableIndex = VarInt.decodeInt(bis);
+    String tableName = tables.get(tableIndex);
+
+    ObjectInputStream in = new ObjectInputStream(bis);
+    KeySet keySet;
+    try {
+      keySet = (KeySet) in.readObject();
+    } catch (ClassNotFoundException e) {
+      throw new RuntimeException(e);
+    }
+    return Mutation.delete(tableName, keySet);
+  }
+
+  private void encodeModification(ByteArrayOutputStream bos, Mutation m) throws IOException {
+    List<String> allTables = schema.getTables();
+    String tableName = m.getTable().toLowerCase();
+    int tableIndex = allTables.indexOf(tableName);
+    VarInt.encode(tableIndex, bos);
+    List<SpannerSchema.Column> columns = schema.getColumns(tableName);
+    checkArgument(columns != null, "Schema for table " + tableName + " not " + "found");
+    Map<String, Value> map = mutationAsMap(m);
+    // java.util.BitSet#toByteArray returns array of unpredictable length. Using byte arrays
+    // instead.
+    int bitsetSize = (columns.size() + 7) / 8;
+    byte[] exists = new byte[bitsetSize];
+    byte[] nulls = new byte[bitsetSize];
+    for (int i = 0; i < columns.size(); i++) {
+      String columnName = columns.get(i).getName();
+      boolean columnExists = map.containsKey(columnName);
+      boolean columnNull = columnExists && map.get(columnName).isNull();
+      if (columnExists) {
+        setBit(exists, i);
+      }
+      if (columnNull) {
+        setBit(nulls, i);
+        map.remove(columnName);
+      }
+    }
+    bos.write(exists);
+    bos.write(nulls);
+    for (int i = 0; i < columns.size(); i++) {
+      if (!getBit(exists, i) || getBit(nulls, i)) {
+        continue;
+      }
+      SpannerSchema.Column column = columns.get(i);
+      Value value = map.remove(column.getName());
+      encodeValue(bos, value);
+    }
+    checkArgument(map.isEmpty(),
+        "Columns " + map.keySet() + " were not defined in" + " table " + m.getTable());
+  }
+
+  private void encodeValue(ByteArrayOutputStream bos, Value value) throws IOException {
+    switch (value.getType().getCode()) {
+      case ARRAY:
+        encodeArray(bos, value);
+        break;
+      default:
+        encodePrimitive(bos, value);
+    }
+  }
+
+  private void encodeArray(ByteArrayOutputStream bos, Value value) throws IOException {
+    // TODO: avoid using Java serialization here.
+    ObjectOutputStream out = new ObjectOutputStream(bos);
+    switch (value.getType().getArrayElementType().getCode()) {
+      case BOOL: {
+        out.writeObject(new ArrayList<>(value.getBoolArray()));
```

Comment:
* Why wrap the primitive arrays into an ArrayList, why not just writeObject(value.getBoolArray())?
* I suppose value.getBoolArray() is probably extremely inefficient both in terms of performance and in terms of serialized size, as are other methods for encoding primitive arrays below as well. Why do you need this hand-encoding at all? Can you just use value.valueToProto() and use the proto's wire format?